app/src/main/kotlin/compiler/semantic_analysis/NameResolver.kt

```diff
@@ -1,12 +1,269 @@
+
+                    for (param in node.parameters) { // and then add their names
+                        reportIfNameConflict(param.name, newScope) // verifies that the parameters have different names
+                        addName(param.name, param, newScope)
+                    }
+
+                    node.body.forEach { analyzeNode(it, newScope) }
+                    destroyScope(newScope) // forget about parameters and names introduced in the body
+                }
+
+                is Function.Parameter -> {
+                    node.defaultValue?.let { analyzeNode(it, currentScope) }
+                }
+
+                // Expressions
+
+                is Expression.Variable -> {
+                    reportIfVariableUndefined(node.name)
+                    reportIfFunctionUsedAsVariable(node.name)
+                    nameDefinitions[node] = visibleNames[node.name]!!.topVariable()
+                }
+
+                is Expression.FunctionCall -> {
+                    reportIfFunctionUndefined(node.name)
+                    reportIfVariableUsedAsFunction(node.name)
+                    nameDefinitions[node] = visibleNames[node.name]!!.topFunction()
+                    node.arguments.forEach { analyzeNode(it, currentScope) }
+                }
+
+                is Expression.FunctionCall.Argument -> {
+                    analyzeNode(node.value, currentScope)
+                }
+
+                is Expression.UnaryOperation -> {
+                    analyzeNode(node.operand, currentScope)
+                }
+
+                is Expression.BinaryOperation -> {
+                    analyzeNode(node.leftOperand, currentScope)
+                    analyzeNode(node.rightOperand, currentScope)
+                }
+
+                is Expression.Conditional -> {
+                    analyzeNode(node.condition, currentScope)
+                    analyzeNode(node.resultWhenTrue, currentScope)
+                    analyzeNode(node.resultWhenFalse, currentScope)
+                }
+
+                // Statements
+
+                is Statement.Evaluation -> {
+                    analyzeNode(node.expression, currentScope)
+                }
+
+                is Statement.VariableDefinition -> {
+                    reportIfNameConflict(node.variable.name, currentScope)
+
+                    // first analyze the variable, then add name, because we can't have self-referencing definitions
+                    analyzeNode(node.variable, currentScope)
+                    addName(node.variable.name, node.variable, currentScope)
+                }
+
+                is Statement.FunctionDefinition -> {
+                    reportIfNameConflict(node.function.name, currentScope)
+
+                    // first add name, then analyze because we can have recursive calls in the body
+                    addName(node.function.name, node.function, currentScope)
+                    analyzeNode(node.function, currentScope)
```

Comment:
Cases `Statement.FunctionDefinition` and `Global.FunctionDefinition` are basically the same - could we move the repeated code to the `Function` case and here simply write `is Statement.FunctionDefinition -> analyzeNode(node.function, currentScope)`? Could we do the same for `Statement.VariableDefinition`/`Global.VariableDefinition`?